packages/assets-controllers/src/MultichainBalancesController/MultichainBalancesController.ts

```diff
@@ -0,0 +1,396 @@
+   *
+   * @param accountId - The account ID.
+   */
+  async updateBalance(accountId: string) {
+    // NOTE: No need to track the account here, since we start tracking those when
+    // the "AccountsController:accountAdded" is fired.
+    await this.#tracker.updateBalance(accountId);
+  }
+
+  /**
+   * Updates the balances of all supported accounts. This method doesn't return
+   * anything, but it updates the state of the controller.
+   */
+  async updateBalances() {
+    await this.#tracker.updateBalances();
+  }
+
+  /**
+   * Checks for non-EVM accounts.
+   *
+   * @param account - The new account to be checked.
+   * @returns True if the account is a non-EVM account, false otherwise.
+   */
+  #isNonEvmAccount(account: InternalAccount): boolean {
+    return (
+      !isEvmAccountType(account.type) &&
+      // Non-EVM accounts are backed by a Snap for now
+      account.metadata.snap !== undefined
+    );
+  }
+
+  /**
+   * Handles changes when a new account has been added.
+   *
+   * @param account - The new account being added.
+   */
+  async #handleOnAccountAdded(account: InternalAccount) {
+    this.#trackAccount(account);
+    // NOTE: Unfortunately, we cannot update the balance right away here, because
+    // messenger's events are running synchronously and fetching the balance is
+    // asynchronous.
+    // Updating the balance here would resume at some point but the event emitter
+    // will not `await` this (so we have no real control "when" the balance will
+    // really be updated), see:
+    // - https://github.com/MetaMask/core/blob/v213.0.0/packages/accounts-controller/src/AccountsController.ts#L1036-L1039
+  }
+
+  /**
+   * Handles changes when a new account has been removed.
+   *
+   * @param accountId - The account ID being removed.
+   */
+  async #handleOnAccountRemoved(accountId: string) {
+    if (this.#tracker.isTracked(accountId)) {
+      this.#tracker.untrack(accountId);
+    }
+
+    if (accountId in this.state.balances) {
+      this.update((state: Draft<MultichainBalancesControllerState>) => {
+        delete state.balances[accountId];
+        return state;
+      });
+    }
+  }
+
+  /**
+   * Get the balances for an account.
+   *
+   * @param accountId - ID of the account to get balances for.
+   * @param snapId - ID of the Snap which manages the account.
+   * @param assetTypes - Array of asset types to get balances for.
+   * @returns A map of asset types to balances.
+   */
+  async #getBalances(
+    accountId: string,
+    snapId: string,
+    assetTypes: CaipAssetType[],
+  ): Promise<Record<CaipAssetType, Balance>> {
+    return await this.#getClient(snapId).getAccountBalances(
+      accountId,
+      assetTypes,
+    );
+  }
+
+  /**
+   * Gets a `KeyringClient` for a Snap.
+   *
+   * @param snapId - ID of the Snap to get the client for.
+   * @returns A `KeyringClient` for the Snap.
+   */
+  #getClient(snapId: string): KeyringClient {
+    return new KeyringClient({
```

Comment:
I see we are eagerly creating a client per snapId. What if we cached these clients in a map on the class. The snapId would be the key and the client the value. I am not sure if there would be any side effects with state management of the clients but it could reduce memory use and computation. The code would look something like this...

```ts
#keyringClients: Map<string, KeyringClient> = new Map();

# getClient(snapId: string): KeyringClient {
  if (!this.#keyringClients.has(snapId)) {
    const client = new KeyringClient({
      send: async (request: JsonRpcRequest) =>
        (await this.messagingSystem.call('SnapController:handleRequest', {
          snapId: snapId as SnapId,
          origin: 'metamask',
          handler: HandlerType.OnKeyringRequest,
          request,
        })) as Promise<Json>,
    });
    this.#keyringClients.set(snapId, client);
  }
  return this.#keyringClients.get(snapId)!;
}
```

Since we will likely not have many snaps, this is a small optimization that is not a blocker for this PR.

Reply:
I do agree that this would improve performance but not really in memory since now we're using it to keep a map of keyring clients, this is a good suggestion, let me think a bit more about it

Reply:
IMO that's a good suggestion yes. But we could having the client creation happening when we receive "accounts events" (like added/removed or upon initial tracking), so that `#getClient` would not implement this in a lazy-way (by lazy I mean, lazy/delayed instantiation here)

But this can probably wait, cause I don't expect this to be impacting performance that much TBH